<!DOCTYPE html>
  <head>

 <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta name="description" content="">
<meta name="keywords" content=" ">
<title>Getting Started  | Manuals</title>
<link rel="stylesheet" type="text/css" href="../css/syntax.css">
<link rel="stylesheet" type="text/css" href="../css/font-awesome.min.css">
<!--<link rel="stylesheet" type="text/css" href="../css/bootstrap.min.css">-->
<link rel="stylesheet" type="text/css" href="../css/modern-business.css">
<link rel="stylesheet" type="text/css" href="../css/lavish-bootstrap.css">
<link rel="stylesheet" type="text/css" href="../css/customstyles.css">
<link rel="stylesheet" type="text/css" href="../css/theme-blue.css">
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-cookie/1.4.1/jquery.cookie.min.js"></script>
<script src="../js/jquery.navgoco.min.js"></script>
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/js/bootstrap.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/2.0.0/anchor.min.js"></script>
<script src="../js/toc.js"></script>
<script src="../js/customscripts.js"></script>
<link rel="shortcut icon" href="../common_images/favicon.ico" type="image/x-icon">
<!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
<!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
<script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
<![endif]-->







 

<script>
  $(function () {
      $('[data-toggle="tooltip"]').tooltip()
  })
</script>



  </head>

<body>

   <!-- Navigation -->
<nav class="navbar navbar-inverse navbar-fixed-top" role="navigation">
    <div class="container topnavlinks">
        <div class="navbar-header">
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>

            <a class="fa fa-home fa-lg navbar-brand" href="home.html">&nbsp;<span class="projectTitle"> Data Analysis Tutorials</span></a>

        </div>
        <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
            <ul class="nav navbar-nav navbar-right">
                <!-- entries without drop-downs appear here -->
                <!-- conditional logic to control which topnav appears for the audience defined in the configuration file.-->
                



















                
                
                
                
                <li><a href="../mydoc/search.html">Search</a></li>
                
                
                
                
                
                <li><a href="https://github.com/tgirke/manuals" target="_blank">Github Repo</a></li>
                
                
                
                


                <!-- entries with drop-downs appear here -->
                <!-- conditional logic to control which topnav appears for the audience defined in the configuration file.-->

                <li class="dropdown">
                    
                    
                    
                    <a href="#" class="dropdown-toggle" data-toggle="dropdown">Resources<b class="caret"></b></a>
                    <ul class="dropdown-menu">
                        
                        
                        
                        <li><a href="https://cran.r-project.org/" target="_blank">CRAN</a></li>
                        
                        
                        
                        
                        
                        <li><a href="http://bioconductor.org/" target="_blank">Bioconductor</a></li>
                        
                        
                        
                        
                        
                        <li><a href="https://guides.github.com/activities/hello-world/" target="_blank">GitHub</a></li>
                        
                        
                        
                        
                        
                        <li><a href="http://jekyllrb.com/" target="_blank">Jekyll</a></li>
                        
                        
                        
                        
                        
                        <li><a href="http://girke.bioinformatics.ucr.edu/" target="_blank">Girke Lab</a></li>
                        
                        
                        
                        
                        
                        <li><a href="http://idratherbewriting.com/documentation-theme-jekyll/mydoc/home.html" target="_blank">Jekyll Doc Theme</a></li>
                        
                        
                        
                        

                    </ul>
                </li>
                
                


                <!-- special insertion -->

               
                
                
                <!-- Send feedback function -->
<script>
function SendLinkByMail(href) {
var subject= "Manuals feedback";
var body = "I have some feedback about the Getting Started page: ";
body += window.location.href;
body += "";
var uri = "mailto:?subject=";
uri += encodeURIComponent(subject);
uri += "&body=";
uri += encodeURIComponent(body);
window.location.href = uri;
}
</script>

<li><a href="javascript:(function()%7BSendLinkByMail()%3B%7D)()%3B" target="_blank"><i class="fa fa-envelope-o"></i> Feedback</a></li>


                <!--uncomment this block if you want simple search instead of algolia-->
                <li>
                     <!--start search-->
                    <div id="search-demo-container">
                        <input type="text" id="search-input" placeholder="search...">
                        <ul id="results-container"></ul>
                    </div>
                    <script src="../js/jekyll-search.js" type="text/javascript"></script>
                    <script type="text/javascript">
                        SimpleJekyllSearch.init({
                            searchInput: document.getElementById('search-input'),
                            resultsContainer: document.getElementById('results-container'),
                            dataSource: '../search.json',
                            searchResultTemplate: '<li><a href="{url}" title="Getting Started">{title}</a></li>',
                        noResultsText: 'No results found.',
                                limit: 10,
                                fuzzy: true,
                        })
                    </script>
                     <!--end search-->
                </li>

            
        </div>
        <!-- /.container -->
</nav>


    <!-- Page Content -->
    <div class="container">
        <div class="col-lg-12">&nbsp;</div>


<!-- Content Row -->
<div class="row">
    <!-- Sidebar Column -->
    <div class="col-md-3">

        <script>

            $(document).ready(function() {
                // Initialize navgoco with default options
                $("#mysidebar").navgoco({
                    caretHtml: '',
                    accordion: true,
                    openClass: 'active', // open
                    save: false, // leave false or nav highlighting doesn't work right
                    cookie: {
                        name: 'navgoco',
                        expires: false,
                        path: '/'
                    },
                    slide: {
                        duration: 400,
                        easing: 'swing'
                    }
                });

                $("#collapseAll").click(function(e) {
                    e.preventDefault();
                    $("#mysidebar").navgoco('toggle', false);
                });

                $("#expandAll").click(function(e) {
                    e.preventDefault();
                    $("#mysidebar").navgoco('toggle', true);
                });

            });

        </script>


        



















        <ul id="mysidebar" class="nav">

            <span class="siteTagline">Manuals</span>
            <span class="versionTagline">Under Construction!</span>

            
            
            
                
            
            <li><a href="#">Overview</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/home.html">Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_classes.html">Class teaching</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Introduction to R</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_1.html">1. Overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_2.html">2. R Package Repositories</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_3.html">3. Installation of R and Add-on Packages</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_4.html">4. Getting Around</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_5.html">5. Basic Syntax</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_6.html">6. Data Types</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_7.html">7. Data objects</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_8.html">8. Graphics example</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_9.html">9. Session Info</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">ChemmineR</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_01.html">1. Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_02.html">2. Getting Started</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_03.html">3. OpenBabel Functions</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_04.html">4. Overview of Classes and Functions</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_05.html">5. Import of Compounds</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_06.html">6. Export of Compounds</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_07.html">7. Format Interconversions</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_08.html">8. Splitting SD Files</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_09.html">9. Streaming Through Large SD Files</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_10.html">10. Storing Compounds in an SQL Database</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_11.html">11. Working with SDF/SDFset Classes</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_12.html">12. Molecular Property Functions (Physicochemical Descriptors)</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_13.html">13. Bond Matrices</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_14.html">14. Charges and Missing Hydrogens</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_15.html">15. Ring Perception and Aromaticity Assignment</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_16.html">16. Rendering Chemical Structure Images</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_17.html">17. Similarity Comparisons and Searching</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_18.html">18. Clustering</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_19.html">19. Searching PubChem</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_20.html">20. ChemMine Tools R Interface</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_21.html">21. Version Information</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_22.html">22. Funding</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_ChemmineR_23.html">23. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">bioassayR</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_bioassayR_1.html">1. Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_bioassayR_2.html">2. Getting Started</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_bioassayR_3.html">3. Examples</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_bioassayR_4.html">4. Building PubChem BioAssay Database</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_bioassayR_5.html">5. Version Information</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_bioassayR_6.html">6. Funding</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_bioassayR_7.html">7. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">systemPipeR</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeR_1.html">1. Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li class="active"><a href="../mydoc/mydoc_systemPipeR_2.html">2. Getting Started</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeR_3.html">3. Workflow overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeR_4.html">4. Workflow templates</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeR_5.html">5. Version information</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeR_6.html">6. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Internal notes</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_internal_notes.html">Administration of site</a></li>
                    

                    

                    
                    
                </ul>
                
                
                


                <!-- if you aren't using the accordion, uncomment this block:

                     <p class="external">
                         <a href="#" id="collapseAll">Collapse All</a> | <a href="#" id="expandAll">Expand All</a>
                     </p>
                 -->

    </div>
    <!-- this highlights the active parent class in the navgoco sidebar. this is critical so that the parent expands when you're viewing a page. This must appear below the sidebar code above. Otherwise, if placed inside customscripts.js, the script runs before the sidebar code runs and the class never gets inserted.-->
    <script>$("li.active").parents('li').toggleClass("active");</script>


            <!-- Content Column -->
            <div class="col-md-9">
                
                <div class="post-header">
   <h1 class="post-title-main">Getting Started</h1>
</div>

<div class="post-content">

   

<!-- this handles the automatic toc. use ## for subheads to auto-generate the on-page minitoc. if you use html tags, you must supply an ID for the heading element in order for it to appear in the minitoc. -->
<script>
$( document ).ready(function() {
  // Handler for .ready() called.

$('#toc').toc({ minimumHeaders: 0, listType: 'ul', showSpeed: 0, headers: 'h2,h3,h4' });

/* this offset helps account for the space taken up by the floating toolbar. */
$('#toc').on('click', 'a', function() {
  var target = $(this.getAttribute('href'))
    , scroll_target = target.offset().top

  $(window).scrollTop(scroll_target - 10);
  return false
})
  
});
</script>


<div id="toc"></div>


<!--
    
    <a target="_blank" href="https://github.com/tgirke/manuals/tree/master/mydoc/" class="btn btn-default githubEditButton" role="button"><i class="fa fa-github fa-lg"></i> Edit me</a>
    
-->
<h2 id="installation">Installation</h2>

<p>The R software for running <a href="http://www.bioconductor.org/packages/devel/bioc/html/systemPipeR.html"><em><code>systemPipeR</code></em></a> can be downloaded from <a href="http://cran.at.r-project.org/"><em>CRAN</em></a>. The <em><code>systemPipeR</code></em> environment can be installed from R using the <em><code>biocLite</code></em> install command. The associated data package <a href="https://github.com/tgirke/systemPipeRdata"><em><code>systemPipeRdata</code></em></a> can be used to generate <em><code>systemPipeR</code></em> workflow environments with a single command (<a href="#load-sample-data-and-workflow-templates">see below</a>) containing all parameter files and sample data required to quickly test and run workflows.</p>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">source</span><span class="p">(</span><span class="s2">"http://bioconductor.org/biocLite.R"</span><span class="p">)</span> <span class="c1"># Sources the biocLite.R installation script 
</span><span class="n">biocLite</span><span class="p">(</span><span class="s2">"systemPipeR"</span><span class="p">)</span> <span class="c1"># Installs systemPipeR 
</span><span class="n">biocLite</span><span class="p">(</span><span class="s2">"systemPipeRdata"</span><span class="p">)</span> <span class="err">#</span> <span class="n">Installs</span> <span class="n">systemPipeRdata</span></code></pre></div>

<h2 id="loading-package-and-documentation">Loading package and documentation</h2>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">library</span><span class="p">(</span><span class="s2">"systemPipeR"</span><span class="p">)</span> <span class="c1"># Loads the package
</span><span class="n">library</span><span class="p">(</span><span class="n">help</span><span class="o">=</span><span class="s2">"systemPipeR"</span><span class="p">)</span> <span class="c1"># Lists package info
</span><span class="n">vignette</span><span class="p">(</span><span class="s2">"systemPipeR"</span><span class="p">)</span> <span class="err">#</span> <span class="n">Opens</span> <span class="n">vignette</span></code></pre></div>

<h2 id="load-sample-data-and-workflow-templates">Load sample data and workflow templates</h2>
<p>The mini sample FASTQ files used by this overview vignette as well as the associated workflow reporting vignettes can be loaded via the <em><code>systemPipeRdata</code></em> package as shown below. The chosen data set <a href="http://www.ncbi.nlm.nih.gov/sra/?term=SRP010938"><code>SRP010938</code></a> contains 18 paired-end (PE) read sets from <em>Arabidposis thaliana</em> [@Howard2013-fq]. To minimize processing time during testing, each FASTQ file has been subsetted to 90,000-100,000 randomly sampled PE reads that map to the first 100,000 nucleotides of each chromosome of the <em>A. thalina</em> genome. The corresponding reference genome sequence (FASTA) and its GFF annotion files (provided in the same download) have been truncated accordingly. This way the entire test sample data set requires less than 200MB disk storage space. A PE read set has been chosen for this test data set for flexibility, because it can be used for testing both types of analysis routines requiring either SE (single end) reads or PE reads.</p>

<p>The following loads one of the available NGS workflow templates (here RNA-Seq) into the user’s current working directory. At the moment, the package includes workflow templates for RNA-Seq, ChIP-Seq, VAR-Seq and Ribo-Seq. Templates for additional NGS applications will be provided in the future.</p>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">library</span><span class="p">(</span><span class="n">systemPipeRdata</span><span class="p">)</span>
<span class="n">genWorkenvir</span><span class="p">(</span><span class="n">workflow</span><span class="o">=</span><span class="s2">"rnaseq"</span><span class="p">)</span>
<span class="n">setwd</span><span class="p">(</span><span class="s2">"rnaseq"</span><span class="p">)</span></code></pre></div>

<p>The working environment of the sample data loaded in previous step contains the following preconfigured directory structure:</p>

<ul>
  <li><span style="color:grey"><em><strong>workflow/</strong></em></span>
    <ul>
      <li>This is the directory of the R session running the workflow.</li>
      <li>Run script ( <em>*.Rnw</em> or <em>*.Rmd</em>) and sample annotation (<em>targets.txt</em>) files are located here.</li>
      <li>Note, this directory can have any name (<em>e.g.</em> <span style="color:grey"><em><strong>rnaseq</strong></em></span>, <span style="color:grey"><em><strong>varseq</strong></em></span>). Changing its name does not require any modifications in the run script(s).</li>
      <li>Important subdirectories:
        <ul>
          <li><span style="color:grey"><em><strong>param/</strong></em></span>
            <ul>
              <li>Stores parameter files such as: <em>*.param</em>, <em>*.tmpl</em> and <em>*_run.sh</em>.</li>
            </ul>
          </li>
          <li><span style="color:grey"><em><strong>data/</strong></em> </span>
            <ul>
              <li>FASTQ samples</li>
              <li>Reference FASTA file</li>
              <li>Annotations</li>
              <li>etc.</li>
            </ul>
          </li>
          <li><span style="color:grey"><em><strong>results/</strong></em></span>
            <ul>
              <li>Alignment, variant and peak files (BAM, VCF, BED)</li>
              <li>Tabular result files</li>
              <li>Images and plots</li>
              <li>etc.</li>
            </ul>
          </li>
        </ul>
      </li>
    </ul>
  </li>
</ul>

<p>The sample workflows provided by the package are based on the above directory structure, where directory names are indicated in  <span style="color:grey"><em><strong>grey</strong></em></span>. 
Users can change this structure as needed, but need to adjust the code in their workflows accordingly.</p>

<p>The following parameter files are included in each workflow template:</p>

<ol>
  <li><em><code>targets.txt</code></em>: initial one provided by user; downstream <em><code>targets_*.txt</code></em> files are generated automatically</li>
  <li><em><code>*.param</code></em>: defines parameter for input/output file operations, <em>e.g.</em> <em><code>trim.param</code></em>, <em><code>bwa.param</code></em>, <em><code>vartools.parm</code></em>, …</li>
  <li><em><code>*_run.sh</code></em>: optional bash script, <em>e.g.</em>: <em><code>gatk_run.sh</code></em></li>
  <li>Compute cluster environment (skip on single machine):
    <ul>
      <li><em><code>.BatchJobs</code></em>: defines type of scheduler for <em><code>BatchJobs</code></em></li>
      <li><em><code>*.tmpl</code></em>: specifies parameters of scheduler used by a system, <em>e.g.</em> Torque, SGE, StarCluster, Slurm, etc.</li>
    </ul>
  </li>
</ol>

<h2 id="structure-of-targets-file">Structure of <em><code>targets</code></em> file</h2>

<p>The <em><code>targets</code></em> file defines all input files (<em>e.g.</em> FASTQ, BAM, BCF) and sample comparisons of an analysis workflow. The following shows the format of a sample <em><code>targets</code></em> file included in the package. It also can be viewed and downloaded from <em><code>systemPipeR</code></em>’s GitHub repository <a href="https://github.com/tgirke/systemPipeR/blob/master/inst/extdata/targets.txt">here</a>. In a target file with a single type of input files, here FASTQ files of single end (SE) reads, the first three columns are mandatory including their column names, while it is four mandatory columns for FASTQ files of PE reads. All subsequent columns are optional and any number of additional columns can be added as needed.</p>

<h3 id="structure-of-targets-file-for-single-end-se-samples">Structure of <em><code>targets</code></em> file for single end (SE) samples</h3>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">library</span><span class="p">(</span><span class="n">systemPipeR</span><span class="p">)</span>
<span class="n">targetspath</span> <span class="o">&lt;-</span> <span class="n">system.file</span><span class="p">(</span><span class="s2">"extdata"</span><span class="p">,</span> <span class="s2">"targets.txt"</span><span class="p">,</span> <span class="n">package</span><span class="o">=</span><span class="s2">"systemPipeR"</span><span class="p">)</span> 
<span class="n">read.delim</span><span class="p">(</span><span class="n">targetspath</span><span class="p">,</span> <span class="n">comment.char</span> <span class="o">=</span> <span class="s2">"#"</span><span class="p">)</span></code></pre></div>

<div class="highlight"><pre><code class="language-txt" data-lang="txt">##                    FileName SampleName Factor SampleLong Experiment        Date
## 1  ./data/SRR446027_1.fastq        M1A     M1  Mock.1h.A          1 23-Mar-2012
## 2  ./data/SRR446028_1.fastq        M1B     M1  Mock.1h.B          1 23-Mar-2012
## 3  ./data/SRR446029_1.fastq        A1A     A1   Avr.1h.A          1 23-Mar-2012
## 4  ./data/SRR446030_1.fastq        A1B     A1   Avr.1h.B          1 23-Mar-2012
## 5  ./data/SRR446031_1.fastq        V1A     V1   Vir.1h.A          1 23-Mar-2012
## 6  ./data/SRR446032_1.fastq        V1B     V1   Vir.1h.B          1 23-Mar-2012
## 7  ./data/SRR446033_1.fastq        M6A     M6  Mock.6h.A          1 23-Mar-2012
## 8  ./data/SRR446034_1.fastq        M6B     M6  Mock.6h.B          1 23-Mar-2012
## 9  ./data/SRR446035_1.fastq        A6A     A6   Avr.6h.A          1 23-Mar-2012
## 10 ./data/SRR446036_1.fastq        A6B     A6   Avr.6h.B          1 23-Mar-2012
## 11 ./data/SRR446037_1.fastq        V6A     V6   Vir.6h.A          1 23-Mar-2012
## 12 ./data/SRR446038_1.fastq        V6B     V6   Vir.6h.B          1 23-Mar-2012
## 13 ./data/SRR446039_1.fastq       M12A    M12 Mock.12h.A          1 23-Mar-2012
## 14 ./data/SRR446040_1.fastq       M12B    M12 Mock.12h.B          1 23-Mar-2012
## 15 ./data/SRR446041_1.fastq       A12A    A12  Avr.12h.A          1 23-Mar-2012
## 16 ./data/SRR446042_1.fastq       A12B    A12  Avr.12h.B          1 23-Mar-2012
## 17 ./data/SRR446043_1.fastq       V12A    V12  Vir.12h.A          1 23-Mar-2012
## 18 ./data/SRR446044_1.fastq       V12B    V12  Vir.12h.B          1 23-Mar-2012</code></pre></div>

<p>To work with custom data, users need to generate a <em><code>targets</code></em> file containing the paths to their own FASTQ files and then provide under <em><code>targetspath</code></em> the path to the corresponding <em><code>targets</code></em> file.</p>

<h3 id="structure-of-targets-file-for-paired-end-pe-samples">Structure of <em><code>targets</code></em> file for paired end (PE) samples</h3>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">targetspath</span> <span class="o">&lt;-</span> <span class="n">system.file</span><span class="p">(</span><span class="s2">"extdata"</span><span class="p">,</span> <span class="s2">"targetsPE.txt"</span><span class="p">,</span> <span class="n">package</span><span class="o">=</span><span class="s2">"systemPipeR"</span><span class="p">)</span>
<span class="n">read.delim</span><span class="p">(</span><span class="n">targetspath</span><span class="p">,</span> <span class="n">comment.char</span> <span class="o">=</span> <span class="s2">"#"</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">]</span></code></pre></div>

<div class="highlight"><pre><code class="language-txt" data-lang="txt">##                  FileName1                FileName2 SampleName Factor SampleLong Experiment
## 1 ./data/SRR446027_1.fastq ./data/SRR446027_2.fastq        M1A     M1  Mock.1h.A          1
## 2 ./data/SRR446028_1.fastq ./data/SRR446028_2.fastq        M1B     M1  Mock.1h.B          1</code></pre></div>

<h3 id="sample-comparisons">Sample comparisons</h3>
<p>Sample comparisons are defined in the header lines of the <em><code>targets</code></em> file starting with ‘<code># &lt;CMP&gt;</code>’.</p>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">readLines</span><span class="p">(</span><span class="n">targetspath</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">]</span></code></pre></div>

<div class="highlight"><pre><code class="language-txt" data-lang="txt">## [1] "# Project ID: Arabidopsis - Pseudomonas alternative splicing study (SRA: SRP010938; PMID: 24098335)"                                                                              
## [2] "# The following line(s) allow to specify the contrasts needed for comparative analyses, such as DEG identification. All possible comparisons can be specified with 'CMPset: ALL'."
## [3] "# &lt;CMP&gt; CMPset1: M1-A1, M1-V1, A1-V1, M6-A6, M6-V6, A6-V6, M12-A12, M12-V12, A12-V12"                                                                                             
## [4] "# &lt;CMP&gt; CMPset2: ALL"</code></pre></div>

<p>The function <em><code>readComp</code></em> imports the comparison information and stores it in a <em><code>list</code></em>. Alternatively, <em><code>readComp</code></em> can obtain the comparison information from the corresponding <em><code>SYSargs</code></em> object (see below). 
Note, these header lines are optional. They are mainly useful for controlling comparative analyses according to certain biological expectations, such as identifying differentially expressed genes in RNA-Seq experiments based on simple pair-wise comparisons.</p>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">readComp</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="n">targetspath</span><span class="p">,</span> <span class="n">format</span><span class="o">=</span><span class="s2">"vector"</span><span class="p">,</span> <span class="n">delim</span><span class="o">=</span><span class="s2">"-"</span><span class="p">)</span></code></pre></div>

<div class="highlight"><pre><code class="language-txt" data-lang="txt">## $CMPset1
## [1] "M1-A1"   "M1-V1"   "A1-V1"   "M6-A6"   "M6-V6"   "A6-V6"   "M12-A12" "M12-V12" "A12-V12"
## 
## $CMPset2
##  [1] "M1-A1"   "M1-V1"   "M1-M6"   "M1-A6"   "M1-V6"   "M1-M12"  "M1-A12"  "M1-V12"  "A1-V1"  
## [10] "A1-M6"   "A1-A6"   "A1-V6"   "A1-M12"  "A1-A12"  "A1-V12"  "V1-M6"   "V1-A6"   "V1-V6"  
## [19] "V1-M12"  "V1-A12"  "V1-V12"  "M6-A6"   "M6-V6"   "M6-M12"  "M6-A12"  "M6-V12"  "A6-V6"  
## [28] "A6-M12"  "A6-A12"  "A6-V12"  "V6-M12"  "V6-A12"  "V6-V12"  "M12-A12" "M12-V12" "A12-V12"</code></pre></div>

<h2 id="structure-of-param-file-and-sysargs-container">Structure of <em><code>param</code></em> file and <em><code>SYSargs</code></em> container</h2>
<p>The <em><code>param</code></em> file defines the parameters of a chosen command-line software. The following shows the format of a sample <em><code>param</code></em> file provided by this package.</p>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">parampath</span> <span class="o">&lt;-</span> <span class="n">system.file</span><span class="p">(</span><span class="s2">"extdata"</span><span class="p">,</span> <span class="s2">"tophat.param"</span><span class="p">,</span> <span class="n">package</span><span class="o">=</span><span class="s2">"systemPipeR"</span><span class="p">)</span>
<span class="n">read.delim</span><span class="p">(</span><span class="n">parampath</span><span class="p">,</span> <span class="n">comment.char</span> <span class="o">=</span> <span class="s2">"#"</span><span class="p">)</span></code></pre></div>

<div class="highlight"><pre><code class="language-txt" data-lang="txt">##      PairSet         Name                                  Value
## 1    modules         &lt;NA&gt;                          bowtie2/2.2.5
## 2    modules         &lt;NA&gt;                          tophat/2.0.14
## 3   software         &lt;NA&gt;                                 tophat
## 4      cores           -p                                      4
## 5      other         &lt;NA&gt; -g 1 --segment-length 25 -i 30 -I 3000
## 6   outfile1           -o                            &lt;FileName1&gt;
## 7   outfile1         path                             ./results/
## 8   outfile1       remove                                   &lt;NA&gt;
## 9   outfile1       append                                .tophat
## 10  outfile1 outextension              .tophat/accepted_hits.bam
## 11 reference         &lt;NA&gt;                    ./data/tair10.fasta
## 12   infile1         &lt;NA&gt;                            &lt;FileName1&gt;
## 13   infile1         path                                   &lt;NA&gt;
## 14   infile2         &lt;NA&gt;                            &lt;FileName2&gt;
## 15   infile2         path                                   &lt;NA&gt;</code></pre></div>

<p>The <em><code>systemArgs</code></em> function imports the definitions of both the <em><code>param</code></em> file and the <em><code>targets</code></em> file, and stores all relevant information in a <em><code>SYSargs</code></em> S4 class object. To run the pipeline without command-line software, one can assign <em><code>NULL</code></em> to <em><code>sysma</code></em> instead of a <em><code>param</code></em> file. In addition, one can start the <em><code>systemPipeR</code></em> workflow with pre-generated BAM files by providing a targets file where the <em><code>FileName</code></em> column gives the paths to the BAM files and <em><code>sysma</code></em> is assigned <em><code>NULL</code></em>.</p>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">args</span> <span class="o">&lt;-</span> <span class="n">suppressWarnings</span><span class="p">(</span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="n">parampath</span><span class="p">,</span> <span class="n">mytargets</span><span class="o">=</span><span class="n">targetspath</span><span class="p">))</span>
<span class="n">args</span></code></pre></div>

<div class="highlight"><pre><code class="language-txt" data-lang="txt">## An instance of 'SYSargs' for running 'tophat' on 18 samples</code></pre></div>

<p>Several accessor functions are available that are named after the slot names of the <em><code>SYSargs</code></em> object.</p>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">names</span><span class="p">(</span><span class="n">args</span><span class="p">)</span></code></pre></div>

<div class="highlight"><pre><code class="language-txt" data-lang="txt">##  [1] "targetsin"     "targetsout"    "targetsheader" "modules"       "software"      "cores"        
##  [7] "other"         "reference"     "results"       "infile1"       "infile2"       "outfile1"     
## [13] "sysargs"       "outpaths"</code></pre></div>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">modules</span><span class="p">(</span><span class="n">args</span><span class="p">)</span></code></pre></div>

<div class="highlight"><pre><code class="language-txt" data-lang="txt">## [1] "bowtie2/2.2.5" "tophat/2.0.14"</code></pre></div>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">cores</span><span class="p">(</span><span class="n">args</span><span class="p">)</span></code></pre></div>

<div class="highlight"><pre><code class="language-txt" data-lang="txt">## [1] 4</code></pre></div>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">outpaths</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="m">1</span><span class="p">]</span></code></pre></div>

<div class="highlight"><pre><code class="language-txt" data-lang="txt">##                                                                                                              M1A 
## "/home/tgirke/Dropbox/Websites/manuals/vignettes/systemPipeR/results/SRR446027_1.fastq.tophat/accepted_hits.bam"</code></pre></div>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">sysargs</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="m">1</span><span class="p">]</span></code></pre></div>

<div class="highlight"><pre><code class="language-txt" data-lang="txt">##                                                                                                                                                                                                                                                                                  M1A 
## "tophat -p 4 -g 1 --segment-length 25 -i 30 -I 3000 -o /home/tgirke/Dropbox/Websites/manuals/vignettes/systemPipeR/results/SRR446027_1.fastq.tophat /home/tgirke/Dropbox/Websites/manuals/vignettes/systemPipeR/data/tair10.fasta ./data/SRR446027_1.fastq ./data/SRR446027_2.fastq"</code></pre></div>

<p>The content of the <em><code>param</code></em> file can also be returned as JSON object as follows (requires <em><code>rjson</code></em> package).</p>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="n">parampath</span><span class="p">,</span> <span class="n">mytargets</span><span class="o">=</span><span class="n">targetspath</span><span class="p">,</span> <span class="n">type</span><span class="o">=</span><span class="s2">"json"</span><span class="p">)</span></code></pre></div>

<div class="highlight"><pre><code class="language-txt" data-lang="txt">## [1] "{\"modules\":{\"n1\":\"\",\"v2\":\"bowtie2/2.2.5\",\"n1\":\"\",\"v2\":\"tophat/2.0.14\"},\"software\":{\"n1\":\"\",\"v1\":\"tophat\"},\"cores\":{\"n1\":\"-p\",\"v1\":\"4\"},\"other\":{\"n1\":\"\",\"v1\":\"-g 1 --segment-length 25 -i 30 -I 3000\"},\"outfile1\":{\"n1\":\"-o\",\"v2\":\"&lt;FileName1&gt;\",\"n3\":\"path\",\"v4\":\"./results/\",\"n5\":\"remove\",\"v1\":\"\",\"n2\":\"append\",\"v3\":\".tophat\",\"n4\":\"outextension\",\"v5\":\".tophat/accepted_hits.bam\"},\"reference\":{\"n1\":\"\",\"v1\":\"./data/tair10.fasta\"},\"infile1\":{\"n1\":\"\",\"v2\":\"&lt;FileName1&gt;\",\"n1\":\"path\",\"v2\":\"\"},\"infile2\":{\"n1\":\"\",\"v2\":\"&lt;FileName2&gt;\",\"n1\":\"path\",\"v2\":\"\"}}"</code></pre></div>



<div class="tags">
    
</div>



</div>



<footer>
            <div class="row">
                <div class="col-lg-12 footer">
               &copy;2016 UC Riverside. All rights reserved. <br />
<p>Page last updated:</span> Sat Feb 13 15:31:51 2016<br/> Site last generated: Feb 15, 2016 <br />
<p><img src="../common_images/company_logo.png/></p>
                </div>
            </div>
</footer>

            </div><!-- /.row -->

    </div>    <!-- /.container -->

</body>


</html>

